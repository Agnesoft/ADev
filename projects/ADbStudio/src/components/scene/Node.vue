<template>
    <div class="node" :style="style">
        <Element :element="element" v-on="$listeners" />
    </div>
</template>

<script>
import Element from "./Element"

export default {
    name: "Node",
    components: {
        Element
    },
    props: {
        element: Object,
        viewSize: Object
    },
    data(){
        return {
        }
    },
    computed: {
        style: function(){
            return {
                '--node-size': this.node_size+"rem",
                top: this.top_pos, 
                left: this.left_pos
            }
        },
        top_pos: function(){
            let pos = this.element.coordinates.y*this.viewSize.height;
            return 'calc('+(pos).toString()+"px - "+(this.node_size/2).toString()+"rem)";
        },
        left_pos: function(){
            let pos = this.element.coordinates.x*this.viewSize.width;
            return 'calc('+(pos).toString()+"px - "+(this.node_size/2).toString()+"rem)";
        },
        node_size: function(){
            return "2.4";
        },
    },
}
</script>

<style scoped>
    .node{
        background-color: var(--primary-color);
        color: var(--dark-color);
        position: absolute;
        border-radius: 50%;
        width: var(--node-size);
        height: var(--node-size);
        line-height:  var(--node-size);
        text-align: center;
        cursor: pointer;
        transition: all var(--transition-ease);
        font-weight: 700;
    }
    
    .node:hover{
        box-shadow:  0 0 3px var(--primary-color), 2px 2px 5px var(--dark-color);
    }
    .search-mode .node{
        background-color: #f2ede8;
        color: #c4b1a1;
    }
    .node.selected{
        background-color: var(--primary-color);
        color: var(--dark-color);
    }
    
</style>
