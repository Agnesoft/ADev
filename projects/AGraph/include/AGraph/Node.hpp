// Copyright 2020 Michael Vlach
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef AGRAPH_NODE_HPP
#define AGRAPH_NODE_HPP

#include "AGraphModule.hpp"
#include "Common.hpp"
#include "Element.hpp"

namespace agraph
{
template<typename GraphType, typename GraphImpl>
class Edge;

//! The Node<GraphType, GraphImpl> is a template
//! that represents a node of a graph in given
//! \c GraphType.
//!
//! The node (also known as vertex, point) is an
//! element of a graph. The Node class provides methods
//! to access outgoing and incoming edges and to
//! iterate over them.
//!
//! The basic functionality is provided by the
//! #agraph::Element base class.
template<typename GraphType, typename GraphImpl>
class Node : public Element<Node<GraphType, GraphImpl>, GraphType>
{
public:
    //! Immutable iterator over the node's outgoing
    //! and incoming edges.
    using const_iterator = acore::ForwardIterator<Edge<GraphType, GraphImpl>, Edge<GraphType, GraphImpl>, const Node>;

    //! Default constructor.
    constexpr Node() = default;

    //! Returns the #const_iterator pointing to the
    //! first outgoing edge from this node. It does not
    //! necessarily need to be the first edge created
    //! from this node. If there are no outgoing
    //! edges end() is returned. The node itself
    //! must be valid.
    [[nodiscard]] constexpr auto begin() const -> const_iterator
    {
        const acore::size_type edge = this->graph()->mData.node(this->index()).from;

        if (edge < acore::INVALID_INDEX)
        {
            return const_iterator{edge, this};
        }

        return end();
    }

    //! Returns the combined number of outgoing and
    //! incoming connections from/to this node. This
    //! number may not represent the number of actual
    //! edges if there are circular connections (edges
    //! connecting the node to itself). The node must
    //! be valid.
    [[nodiscard]] constexpr auto edgeCount() const -> acore::size_type
    {
        return fromCount() + toCount();
    }

    //! Returns the #const_iterator pointing to the
    //! invalid edge.
    [[nodiscard]] constexpr auto end() const noexcept -> const_iterator
    {
        return const_iterator{acore::INVALID_INDEX, this};
    }

    //! Returns the number of edges outgoing from
    //! this node. The node must be valid.
    [[nodiscard]] constexpr auto fromCount() const -> acore::size_type
    {
        return this->graph()->mData.node(this->index()).fromCount;
    }

    //! Returns the #const_iterator pointing to the
    //! first incoming edge from this node. It does
    //! not necessarily need to be the first edge
    //! created to this node. If there are no incoming
    //! edges end() is returned. The node must be
    //! valid.
    [[nodiscard]] constexpr auto rbegin() const -> const_iterator
    {
        const acore::size_type edge = this->graph()->mData.node(this->index()).to;

        if (edge < acore::INVALID_INDEX)
        {
            return const_iterator{edgeToIndex(edge), this};
        }

        return rend();
    }

    //! Returns the #const_iterator pointing to the
    //! invalid edge.
    [[nodiscard]] constexpr auto rend() const -> const_iterator
    {
        return end();
    }

    //! Returns the number of edges incoming to
    //! this node. The node must be valid.
    [[nodiscard]] constexpr auto toCount() const -> acore::size_type
    {
        return this->graph()->mData.node(this->index()).toCount;
    }

protected:
    //! Constructs a node with the \a index and
    //! the \a graph.
    constexpr Node(acore::size_type index, const GraphType *graph) noexcept :
        Element<Node<GraphType, GraphImpl>, GraphType>{index, graph}
    {
    }

private:
    friend const_iterator;
    friend GraphImpl;

    //! \private
    //! NEEDED BECAUSE OF THE BUG IN DOXYGEN 1.8.13
    [[nodiscard]] constexpr auto nextIndex(acore::size_type index) const -> acore::size_type
    {
        if (index < acore::INVALID_INDEX)
        {
            return this->graph()->mData.edge(indexToEdge(index)).nextFrom;
        }

        if (acore::INVALID_INDEX < index)
        {
            return edgeToIndex(this->graph()->mData.edge(index).nextTo);
        }

        throw acore::Exception{} << "Cannot incrementing invalid 'agraph::Node::iterator'.";
    }

    [[nodiscard]] constexpr auto referenceAt(acore::size_type index) const -> Edge<GraphType, GraphImpl>
    {
        if (acore::INVALID_INDEX < index)
        {
            return this->graph()->edge(edgeToIndex(index));
        }

        if (index < acore::INVALID_INDEX)
        {
            return this->graph()->edge(index);
        }

        throw acore::Exception{} << "Dereferencing invalid 'agraph::Node::iterator'.";
    }
};
}

#endif
