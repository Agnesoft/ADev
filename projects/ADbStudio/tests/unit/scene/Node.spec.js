import { mount  } from '@vue/test-utils';
import Node from '@/components/scene/Node.vue';

import { localVue } from '../TestUtils'


describe('Node',() => {
    let wrapper
    beforeAll(() => {
        wrapper = mount(Node, {
            localVue,
            propsData: {
                element: {
                    id: 1,
                    coordinates: {
                        x: 0.2,
                        y: 0.5
                    }
                },
                viewSize: {
                    width: 200,
                    height: 100
                }
            },
            computed: {
                node_size() {
                    return "2.4";
                }
              }
          });
    })
    it('renders node id',() => {
        expect(wrapper.text()).toBe("1");
    })
    it('renders node in the correct position', () => {
        let style = wrapper.attributes().style;
        expect(style).toContain('top: calc(50px - 1.2rem)');
        expect(style).toContain('left: calc(40px - 1.2rem)');
    })
    it('renders correctly (snapshot)', () => {
        expect(wrapper.element).toMatchSnapshot()
    })
})
