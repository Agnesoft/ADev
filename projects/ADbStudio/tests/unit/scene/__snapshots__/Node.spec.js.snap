// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node renders correctly (snapshot) 1`] = `
<div
  class="node"
  style="--node-size: 2.4rem; top: calc(50px - 1.2rem); left: calc(40px - 1.2rem);"
>
  <div
    class="element"
  >
    
    1

  </div>
</div>
`;
